gpkit/constraints/set.py

```diff
@@ -0,0 +1,227 @@
+        pass
+
+    def _subconstr_tex(self, excluded=None):
+        "The collapsed appearance of a ConstraintSet"
+        pass
+
+    def flat(self, constraintsets=True):
+        "Yields contained constraints, optionally including constraintsets."
+        for constraint in self:
+            if not isinstance(constraint, ConstraintSet):
+                yield constraint
+            else:
+                if constraintsets:
+                    yield constraint
+                subgenerator = constraint.flat(constraintsets)
+                for yielded_constraint in subgenerator:
+                    yield yielded_constraint
+
+    def sub(self, subs, value=None):
+        "Substitutes in place."
+        if hasattr(self, "cost"):
+            self.cost = self.cost.sub(subs, value)
+        for i, constraint in enumerate(self):
+            self[i] = constraint.sub(subs, value)
+        return self
+
+    @property
+    def varkeys(self):
+        "Varkeys present in the constraints"
+        out = KeySet()
+        if hasattr(self, "cost"):
+            out.update(self.cost.varkeys)
+        for constraint in self:
+            if hasattr(constraint, "varkeys"):
+                out.update(constraint.varkeys)
+        return out
+
+    def as_posyslt1(self):
+        "Returns list of posynomials which must be kept <= 1"
+        posylist, self.posymap = [], []
+        for constraint in self:
+            constraint.substitutions = KeyDict()
+            constraint.substitutions.update(self.substitutions)
+            posys = constraint.as_posyslt1()
+            self.posymap.append(len(posys))
+            posylist.extend(posys)
+        return posylist
+
+    def sens_from_dual(self, p_senss, m_sensss):
+        """Computes constraint and variable sensitivities from dual solution
+
+        Arguments
+        ---------
+        p_senss : list
+            Sensitivity of each posynomial returned by `self.as_posyslt1()`
+
+        m_sensss: list of lists
+            Each posynomial's monomial sensitivities
+
+
+        Returns
+        -------
+        constraint_sens : dict
+            The interesting and computable sensitivities of this constraint
+
+        var_senss : dict
+            The variable sensitivities of this constraint
+        """
+        constr_sens = {}
+        var_senss = HashVector()
+        offset = 0
+        for i, constr in enumerate(self):
+            n_posys = self.posymap[i]
+            p_ss = p_senss[offset:offset+n_posys]
+            m_sss = m_sensss[offset:offset+n_posys]
+            constr_sens[str(constr)], v_ss = constr.sens_from_dual(p_ss, m_sss)
+            var_senss += v_ss
+            offset += n_posys
+        return constr_sens, var_senss
+
+    def as_gpconstr(self, x0):
+        """Returns GPConstraint approximating this constraint at x0
+
+        When x0 is none, may return a default guess."""
+        cs = ConstraintSet([constr.as_gpconstr(x0) for constr in self])
```

Comment:
open to discussion on that. I like the idea that a constraint set is just the recursive object, with no cost for simplicity. we can talk about what a natural inheritance structure under that would look like.